base/usr/share/man/man1/roff.1
.TH ROFF 1 "March, 2026"
.SH NAME
roff \- terrible, hacked\-together RUNOFF implementation
.SH SYNOPSIS
\fBroff\fP [\fB\-W\fP \fIwidth\fP] \fIfile\fP...
.br
\fBroff\fP \fB--help\fP
.SH DESCRIPTION
\fIroff\fP is a terrible, no-good, very-bad, horrendously incomplete implementation
of the \fIRUNOFF\fP text formatter.
.PP
This \fIroff\fP was made entirely to process \fIman\fP pages like this one, and
probably won't work with ones that didn't come with linAIxOS.
.TP
.B \-W \fIwidth\fP
Instead of formatting the output for the width of the terminal, as is the default,
the output will be formatted for a page width of \fIwidth\fP columns.
.SH BUGS
Tons.
.PP
Various macros are not quite correct. Lots of escape sequences are missing.
Tables are not properly supported.
.SH SYNTAX
This \fIroff\fP supports a hacked-together set of macros based on what was used in
.IR nyancat (1)
and
.IR less (1)
(even though we don't have that ported).
.SH MACROS
Lines starting with a \fB.\fP or, alternatively, a \fB'\fP, introduce macros.
.TP
.B .\e"
A comment. The line will be ignored.
.TP
.B .TH \fItopic\fP \fIsection\fP [\fIfooter-middle\fP] [\fIfooter-left\fP] [\fIheader-middle\fP]
Configure the header and footer of the page.
.sp
The \fItopic\fP is the name of this manual, generally the same as the filename prefix, but for
applications it is traditionally capitalized.
.sp
The \fIsection\fP is a numeric section number, 1 through 9, or an alternative string that will
be shown in its place; \fBn\fP is sometimes used by some systems.
.sp
The \fItopic\fP and \fIsection\fP are combined, with the \fItopic\fP underlined, and shown in
the upper left, upper right, and lower right of the output.
.sp
The \fIfooter-middle\fP is usually an authorship date and will be shown in the middle of the
bottom of the output, if present.
.sp
The \fIfooter-left\fP appears at the left of the bottom of the output, if present, but is
usually not included in manual pages.
.sp
The \fIheader-middle\fP, if present, overrides the descriptive name of the \fIsection\fP to
be displayed at middle of the top of the output.
.sp
If the \fIsection\fP is a numeric digit from 0 to 9, and \fIheader-middle\fP is not provided,
the section title will be displayed at the top of the output. Otherwise, \fIsection\fP itself
is displayed.
.sp
 1: General Commands Manual
 2: System Calls Manual
 3: Library Function Names
 4: Special Files
 5: File Formats
 6: Games
 7: Miscellaneous
 8: System Administration Manual
 9: Kernel Programming Manual
.TP
.B .SH \fItitle\fP
Displays a section title. Section titles are traditionally capitalized.
.TP
.B .PP
Regular paragraph break. Sets up standard indentation.
.TP
.B .IP
Indented paragraph. The paragraph is indented more. I'm pretty sure this is the same as \fB.TP\fP?
.TP
.B .TP
Tagged paragraph. The next line will be indented a bit, and then lines after it will be indented more,
I guess. These are all \fB.TP\fP paragraphs, for reference, so they're good for options and definitions.
If the first line and the second line can be placed next to each other, they will be.
.TP
.B .br
Line break. Forces a line feed.
.TP
.B .B
Bold text. The rest of the line will be bold.
.TP
.B .I
Underlined text. Technically, this is supposed to be \fIitalic\fP text, but literally everyone treats
it as underlined instead. There's no standard ROFF thing for actual italic text, which is a shame.
.TP
.B .IR
Do one word with the ita\-er, I mean, underlined font, then switch back to the regular font.
This gets used in some man pages to print the names of manpages, like \fB.IR man (1)\fP or whatever.
.TP
.B .BR
Similar to the above, except with bold instead of underlined. I don't know what you'd want to use this for.
.TP
.B .sp
Insert a vertical space.
.TP
.B .RS
Start a section that has extra indentation. This extra indentation continues until a \fB.RE\fP. This is
supposedly used for block quotes, but it seems to mostly get used for very nested explanations.
.TP
.B .RE
End a "blockquote" section. Restore the normal indentation level.
.TP
.B .nh .hy
These are ignored for compatibility, but they're supposed to toggle hyphenation.
.TP
.B .nf .fi
Between these macros, line feeds are taken literally. There's probably some other stuff
I'm supposed to be doing that I'm not, but whatever.
.TP
.B .TS .TE
These are supposed to introduce tables but just treat them the same as \fB.nf/.fi\fP.
.SH FONT SELECTION
You can use the various macros like \fB.B\fP above to select fonts, but the \fB\\f\fP escape
is a lot more flexible. You can use it to select fonts like this:
.TP
.B \efP
Reset the font to the previous one. There is only one level to the stack, so choose wisely.
.TP
.B \efR
"Roman" \- the default font.
.TP
.B \efB
Bold.
.TP
.B \efI
Supposed to be "italic", but as we all learned earlier it's actually underlined.
.TP
.B \ef(BI
Bold and italic together.
.SH OTHER ESCAPES
.TP
.B \ee
Prints a literal backslash. \e
.TP
.B \e(ha
Prints a caret. \(ha
.TP
.B \e(ti
Prints a tilde. \(ti
.TP
.B \e(aq
Prints a single quote. \(aq
.TP
.B \e(dq
Prints a double quote. \(dq
.TP
.B \e(bu
Prints a list bullet (U+2022). \(bu
.sp
Useful for making lists like this:
 \(bu Apples
 \(bu Bananas
 \(bu Oranges
.TP
.B \e"
Comment. The rest of the line will be ignored. \" If you're reading this manual's source, you can see an example of this here.
.SH SEE ALSO
.IR man (1),
.IR more (1)
.SH HOMEPAGE
.PP
.B linAIxOS:
https://linAIxos.org/
.SH AUTHOR
This \fIroff\fP was written by K. Lange <klange@linAIxos.org>.


apps/roff.c
/**
 * @brief A rudimentary implementation of some parts of ROFF.
 *
 * This is really just intended to read our own man pages, so it
 * only supports functionality that was used in the manpage
 * written for the 'nyancat' utility. It was also tested and
 * expanded based on the manpage for 'less'.
 *
 * @copyright
 * This file is part of linAIxOS and is released under the terms
 * of the NCSA / University of Illinois License - see LICENSE.md
 * Copyright (C) 2026 K. Lange
 */
#include <stdio.h>
#include <stdlib.h>
#include <termios.h>
#include <unistd.h>
#include <string.h>
#include <ctype.h>
#include <errno.h>
#include <sys/ioctl.h>

#define DEFAULT_INDENTATION 7
#define MARGIN_SPACE 5
static struct winsize w;

struct RoffContext {
	char * topic_title;          /* Storage for the current page title, to be displayed in the footer. */
	char * topic_section;        /* Storage for the current section title, to be displayed in the footer. */
	char * topic_date;           /* Storage for the current date string, to be displayed in the footer. */
	int current_x;               /* Display cursor X offset. */
	int indent;                  /* How much we should indent lines before printing. */
	int next_indent;             /* The level of indentation we should switch to after finishing the current line. */
	int extra_indent;            /* An extra amount of indentation we should add to the above values when we see a new paragraph macro. */
	int printing_table;          /* Whether we should handle whitespace more directly. */
	int squish_line;             /* Whether we should try to squish this line into the next one, eg. for a .TP macro. */
	unsigned int previous_font;  /* The previously set font to act as a one-level undo stack. */
	unsigned int current_font;   /* The current font, to be set once we start printing. */
	char * topic_footer;
	char * topic_header;
};


/**
 * @brief Convert a (string) section number into a section description.
 *
 * If the section isn't a single digit, it will be returned directly.
 */
static char * interpret_section(char * section) {
	if (!section) return "Unknown";
	if (strlen(section) == 1 && isdigit(*section)) {
		switch (atoi(section)) {
			case 1: return "General Commands Manual";
			case 2: return "System Calls Manual";
			case 3: return "Library Functions Manual";
			case 4: return "Special Files";
			case 5: return "File Formats";
			case 6: return "Games";
			case 7: return "Miscellaneous";
			case 8: return "System Administration Manual";
			case 9: return "Kernel Programming Manual";
			case 0: return "Section Zero";
		}
	}
	return section; /* Must be custom. */
}

/**
 * @brief Print a formatted title and section number.
 *
 * Eg. foo(1) where 'foo' is underlined.
 */
static void formatted_title_and_section(char * title, char * section) {
	if (!title || !section) {
		printf("??");
		return;
	}
	printf("\033[4m");
	printf("%s", title);
	printf("\033[0m(%s)", section);
}

/**
 * @brief Format and display the title of a page, at the top of the screen.
 *
 * The title and section are displayed in the upper left and right, and the
 * section description is shown in the top center.
 */
static void format_title(struct RoffContext * ctx) {

	char * title = ctx->topic_title;
	char * section = ctx->topic_section;

	/* XXX Should use display width, but too lazy. */
	int title_len = title ? strlen(title) : 0;
	int section_len = section ? strlen(section) : 0;
	char * section_name = ctx->topic_header ? ctx->topic_header : interpret_section(section);
	int section_name_len = strlen(section_name);

	int avail = w.ws_col - 1; /* A gutter is fine. */

	int space_used = (title_len + section_len + 2) * 2 + section_name_len;
	if (avail < space_used) {
		/* Print a short form and hope it works */
		formatted_title_and_section(title, section);
		printf("\n");
		return;
	}

	int space_left = (avail - space_used) / 2;
	int space_right = avail - space_used - space_left;

	formatted_title_and_section(title, section);
	for (int i = 0; i < space_left; ++i) printf(" ");
	printf("%s", section_name);
	for (int i = 0; i < space_right; ++i) printf(" ");
	formatted_title_and_section(title, section);
	printf("\n\n");
}

/**
 * @brief Fromat and display the footer of a page.
 *
 * The "date" is shown in the center, and the title and section are displayed
 * in the lower right.
 */
static void format_footer(struct RoffContext * ctx) {
	int title_len = ctx->topic_title ? strlen(ctx->topic_title) : 0;
	int section_len = ctx->topic_section ? strlen(ctx->topic_section) : 0;
	int date_len = ctx->topic_date ? strlen(ctx->topic_date) : 0;
	int footer_len = ctx->topic_footer ? strlen(ctx->topic_footer) : 0;

	int avail = w.ws_col - 1; /* A gutter is fine. */
	int space_used = title_len + section_len + 2 + date_len + footer_len;

	int space_left = (avail - date_len) / 2 - footer_len;
	int space_right = avail - space_used - space_left;

	printf("%s", ctx->topic_footer ? ctx->topic_footer : "");
	for (int i = 0; i < space_left; ++i) printf(" ");
	printf("%s", ctx->topic_date ? ctx->topic_date : "");
	for (int i = 0; i < space_right; ++i) printf(" ");
	formatted_title_and_section(ctx->topic_title,ctx->topic_section);
	printf("\n");
}

/**
 * @brief Determine how big an escape sequence is.
 *
 * Use this to eat escape sequences from the input when calculating the display width
 * of a word. *len is increased with the number displayed characters, and the total
 * number of input characters processed by the escape is returned.
 *
 * For example, \fI takes three input characters but displays none, and \e takes two
 * input characters and displays one.
 */
static int skip_escape(char *x, size_t *len) {
	switch (x[1]) {
		/* Font selection supports both single and multi-character options */
		case 'f':
			if (x[2] == '(') {
				if (x[3] == 0) return 3;
				if (x[4] == 0) return 4;
				return 5;
			} else if (x[2] == 0) {
				return 2;
			}
			return 3;

		/* Comment */
		case '"': {
			char *c = x;
			while (*c) c++;
			return c - x;
		}

		/* "Italic correction */
		case ',':
		case '/':
			return 2;

		/* Special characters are generally one character, but maybe
		 * some of them will be wider later. */
		case '(':
			(*len)++;
			if (!x[2]) return 2;
			if (!x[3]) return 3;
			return 4;

		/* Assume other escapes print a single character */
		default:
			(*len)++;
			return 2;
	}
}

/**
 * @brief If the cursor is not at the start of a line, print a line feed.
 *
 * @returns 1 if a line feed was printed. Many callers want to then print another one in this case.
 */
static int flush_line(struct RoffContext * ctx) {
	if (ctx->current_x != 0) {
		printf("\033[0m\n");
		ctx->current_x = 0;
		return 1;
	}
	return 0;
}

#define PAIR(a,b) (((unsigned int)a << 8) | (unsigned int)b)

/**
 * @brief Switch the active font.
 *
 * If @c font is 'P', the previous font is retrieved from the undo stack.
 * Otherwise, the current font is stored in the undo stack and the new
 * font is set as the current font.
 *
 * This does not actually change the display font on the terminal; that
 * must be done with @c activate_font() when you are ready to print.
 */
static void switch_font(struct RoffContext * ctx, unsigned int font) {
	if (font == 'P') {
		ctx->current_font = ctx->previous_font;
	} else {
		ctx->previous_font = ctx->current_font;
		ctx->current_font = font;
	}

	/* Aliases first */
	switch (ctx->current_font) {
		case '1': ctx->current_font = 'R'; break;
		case '2': ctx->current_font = 'I'; break;
		case '3': ctx->current_font = 'B'; break;
		case '4': ctx->current_font = PAIR('B','I'); break;
		case PAIR('C','B'): ctx->current_font = 'B'; break;
		case PAIR('C','I'): ctx->current_font = 'I'; break;
		case PAIR('C','R'): ctx->current_font = 'R'; break;
		case PAIR('C','W'): ctx->current_font = 'R'; break;
	}
}

/**
 * @brief Emit the terminal sequence necessary to activate the current font.
 *
 * Prints the escape sequence associated with the current font. Run this
 * before you start printing, but after you've skipped over indentation,
 * as users generally don't want to see a bunch of underlines at the start
 * of a line.
 */
static void activate_font(struct RoffContext * ctx) {
	switch (ctx->current_font) {
		case 'B': printf("\033[0;1m"); break;
		case 'R': printf("\033[0m"); break;
		/* These are actually supposed to be italic, but everyone treats them as underlined (4, rather than 3). */
		case 'I': printf("\033[0;4m"); break;
		case PAIR('B','I'): printf("\033[0;1;4m"); break;
	}
}

/**
 * @brief Process an escape sequence.
 *
 * Actually handle an escape sequence. Switch fonts, print special characters,
 * and so on. As with @c skip_escape, returns the number of input characters
 * handled by the escape.
 *
 * This should always print as many characters (or, really, the width of those
 * characters, when we get around to it) as @c skip_escape said the escape
 * should display when it updated @c len.
 */
static int do_escape(struct RoffContext * ctx, char *x) {
	switch (x[1]) {
		case 'f': /* Font selection takes a few more characters */
			if (x[2] == 0) return 2;
			if (x[2] == '(') {
				if (x[3] == 0) return 3;
				if (x[4] == 0) return 4;
				switch_font(ctx, PAIR(x[3],x[4]));
			} else {
				switch_font(ctx, (unsigned int)x[2]);
			}
			activate_font(ctx);
			return x[2] == '(' ? 5 : 3;
		/* Comment */
		case '"': {
			char *c = x;
			while (*c) c++;
			return c - x;
		}
		case 'e':
			fputc('\\', stdout);
			return 2;
		case ',':
		case '/':
			/* "left and right italic correction"; ignored. */
			return 2;
		case '(':
			if (!x[2]) return 2;
			if (!x[3]) return 3;
			switch (PAIR(x[2],x[3])) {
				case PAIR('h','a'): fputc('^', stdout); break;
				case PAIR('b','u'): printf("•"); break;
				case PAIR('t','i'): fputc('~', stdout); break;
				case PAIR('a','q'): fputc('\'', stdout); break;
				case PAIR('d','q'): fputc('"', stdout); break;
				default: fputc('?', stdout); break;
			}
			return 4;
		default:
			fputc(x[1], stdout);
			return 2;
	}
}

/**
 * @brief Helper to determine if a whole string is spaces.
 *
 * This should probably also handle tabs and maybe some other stuff,
 * or at least just use @c isblank() but it's good enough for now.
 */
static int is_whitespace(char * c) {
	while (*c && *c == ' ') c++;
	if (*c && *c != ' ') return 0;
	return 1;
}

/**
 * @brief Helper to print a set number of spaces.
 *
 * Mostly used to print indentation.
 */
static void print_spaces(int indent) {
	for (int i = 0; i < indent; ++i) fputc(' ', stdout);
}

/**
 * @brief Display a word and handle line wrapping.
 *
 * This is the core of the output formatting.
 *
 * Processes a word from the input text, handling escapes within it,
 * and determines if it will fit on the current line, starting a new
 * line if there is not enough space.
 *
 * In this roff, we don't handle justification or other advanced
 * text layout, so this will always end up immediately printing words.
 */
static size_t process_word(struct RoffContext * ctx, char * c, int delimited) {
	char * c_in = c;
	char * last_word = c;
	size_t last_len = 0;

	/* Collect word */
	while (*c && *c != ' ') {
		if (*c == '\\' && c[1]) {
			c += skip_escape(c, &last_len);
		} else {
			last_len++;
			c++;
		}
	}

	/* Word was empty, done with line. */
	if (!last_len) {
		while (*c && *c == ' ') c++;
		return c - c_in;
	}

	/* Word would wrap, continue to next line and print indentant. */
	if (last_len + ctx->current_x > (size_t)w.ws_col - MARGIN_SPACE) {
		flush_line(ctx);
		print_spaces(ctx->indent);
		ctx->current_x += ctx->indent;
	} else if (ctx->current_x == 0) {
		print_spaces(ctx->indent);
		ctx->current_x += ctx->indent;
	}

	activate_font(ctx);

	/* Print the word */
	char *x = last_word;
	while (*x && x != c) {
		if (*x == '\\' && x[1]) {
			x += do_escape(ctx, x);
		} else {
			fputc(*x, stdout);
			x++;
		}
	}
	ctx->current_x += last_len;

	if (!*c) printf("\033[0m");

	if (ctx->printing_table) {
		while (*c && *c == ' ' && (size_t)ctx->current_x < (size_t)w.ws_col - MARGIN_SPACE) {
			printf(" ");
			ctx->current_x++;
			c++;
		}
	} else {
		if (delimited) {
			fputc(' ', stdout);
			ctx->current_x += 1;
		}
	}

	/* Advance through whitespace */
	while (*c && *c == ' ') c++;

	return c - c_in;
}

#define MAYBE_QUOTES() do { \
	if (*c == '"' && line[len-1] == '"') { \
		line[len-1] = '\0'; \
		len--; \
		c++; \
	} \
} while (0)

static char * collect_arg(char * c, char **out) {
	char * value = NULL;
	*out = NULL;
	while (*c && *c == ' ') c++;
	if (*c) {

		int quoted = 0;
		if (*c == '"') {
			c++;
			quoted = 1;
		}

		value = c;

		if (quoted) {
			while (*c && *c != '"') c++;
		} else {
			while (*c && *c != ' ') c++;
		}
		if (*c) *c++ = '\0';

		/* Skip whitespace */
		while (*c && *c == ' ') c++;
	}

	if (value) *out = strdup(value);

	return c;
}

/**
 * @brief Format and display one file.
 *
 * This is the core of the macro processor.
 *
 * Reads lines from an input file, interprets lines with roff macros,
 * and otherwise processes text for display with @c process_word.
 *
 * @returns 1 on error, 0 if the file was fully processed and displayed.
 */
static int do_file(char ** argv, int i) {
	FILE * f = (!strcmp(argv[i],"-")) ? stdin : fopen(argv[i], "r");
	if (!f) {
		fprintf(stdout, "%s: %s: %s\n", argv[0], argv[i], strerror(errno));
		return 1;
	}

	char * line = NULL;
	size_t avail = 0;
	ssize_t len;

	struct RoffContext ctx = {0};

	ctx.previous_font = 'R';
	ctx.current_font = 'R';

	int ret = 0;

	while ((len = getline(&line, &avail, f)) >= 0) {

		/* Discard linefeed to make this simpler */
		if (len && line[len-1] == '\n') {
			line[len-1] = '\0';
			len--;
		}

		if (!*line && !ctx.printing_table) continue; /* Skip blank lines entirely? Is that correct? */

		if (*line == '\'') line[0] = '.'; /* single tick is an alias for . at the start of a line. */

		/* Skip comment lines */
		if (strstr(line, ".\\\"") == line) continue;

		/* Also skip lines that are just the "control character" */
		if (*line == '.' && is_whitespace(&line[1])) continue;

		char * c = line;
		int delimited = 1;

		if (line[0] == '.') {
			/* Macro directives */

			if (strstr(line,".TH ") == line) {
				if (ctx.topic_title) {
					fprintf(stderr, "%s: %s: More than one .TH found.\n", argv[0], argv[i]);
					ret = 1;
					goto _cleanup;
				}
				/* Topic heading */
				char * c = line + 4;
				c = collect_arg(c, &ctx.topic_title);
				c = collect_arg(c, &ctx.topic_section);
				c = collect_arg(c, &ctx.topic_date);
				c = collect_arg(c, &ctx.topic_footer);
				c = collect_arg(c, &ctx.topic_header);

				format_title(&ctx);
				continue;
			} else if (strstr(line, ".SH ") == line) {
				/* Section heading */
				if (flush_line(&ctx)) printf("\n");
				c = line + 4;
				MAYBE_QUOTES();
				printf("\033[1m%s\033[0m\n", c);
				ctx.indent = ctx.extra_indent + DEFAULT_INDENTATION;
				ctx.next_indent = 0;
				continue;
			} else if (strstr(line, ".PP") == line) {
				/* Paragraph */
				flush_line(&ctx); printf("\n");
				ctx.indent = ctx.extra_indent + DEFAULT_INDENTATION;
				ctx.next_indent = 0;
				continue;
			} else if (strstr(line, ".TP") == line) {
				/* Tagged paragraph */
				if (flush_line(&ctx)) printf("\n");

				ctx.indent = ctx.extra_indent + DEFAULT_INDENTATION;
				ctx.next_indent = ctx.extra_indent + DEFAULT_INDENTATION * 2;
				ctx.squish_line = 1;
				continue;
			} else if (strstr(line, ".br") == line) {
				/* Line break */
				flush_line(&ctx);
				continue;
			} else if (strstr(line, ".B ") == line) {
				/* Bold */
				switch_font(&ctx, 'B');
				c = line + 3;
				MAYBE_QUOTES();
			} else if (strstr(line, ".I ") == line) {
				/* Italic (actually underlined) */
				switch_font(&ctx, 'I');
				c = line + 3;
				MAYBE_QUOTES();
			} else if (strstr(line, ".IR ") == line) {
				/* Italic (actually underlined), then Roman */
				switch_font(&ctx, 'I');
				c = line + 4;
				c += process_word(&ctx, c, 0);
				switch_font(&ctx, 'R');
				/* Continue */
			} else if (strstr(line, ".BR ") == line) {
				/* Bold, then Roman */
				switch_font(&ctx, 'B');
				c = line + 4;
				c += process_word(&ctx, c, 0);
				switch_font(&ctx, 'R');
				/* Continue */
			} else if (strstr(line, ".IP ") == line) {
				/* Indented paragraph */
				if (flush_line(&ctx)) printf("\n");
				ctx.indent = ctx.extra_indent + DEFAULT_INDENTATION;
				ctx.next_indent = ctx.extra_indent + DEFAULT_INDENTATION * 2;
				c = line + 4;
				MAYBE_QUOTES();
				ctx.squish_line = 1;
			} else if (strstr(line, ".IP") == line) {
				/* Indented paragraph but this line is empty */
				if (flush_line(&ctx)) printf("\n");
				ctx.indent = ctx.extra_indent + DEFAULT_INDENTATION;
				ctx.next_indent = ctx.extra_indent + DEFAULT_INDENTATION;
				c = line + 3;
				ctx.squish_line = 1;
			} else if (strstr(line, ".sp") == line) {
				/* Vertical space */
				flush_line(&ctx);
				printf("\n");
				continue;
			} else if (strstr(line, ".RS") == line) {
				/* Block quote start */
				if (flush_line(&ctx)) printf("\n");
				ctx.extra_indent += DEFAULT_INDENTATION;
				ctx.indent += DEFAULT_INDENTATION;
				ctx.next_indent = ctx.indent;
				continue;
			} else if (strstr(line, ".RE") == line) {
				/* Block quote end */
				ctx.indent -= DEFAULT_INDENTATION;
				ctx.next_indent = ctx.indent;
				ctx.extra_indent -= DEFAULT_INDENTATION;
				continue;
			} else if (strstr(line, ".nh") == line) {
				/* TODO disable hyphenation */
				continue;
			} else if (strstr(line, ".hy") == line) {
				/* TODO reenable hyphenation */
				continue;
			} else if (strstr(line, ".nf") == line) {
				/* Handle whitespace, including linebreaks, until .fi */
				flush_line(&ctx);
				ctx.printing_table = 1;
				continue;
			} else if (strstr(line, ".fi") == line) {
				/* End of raw whitespace handling */
				flush_line(&ctx);
				ctx.printing_table = 0;
				continue;
			} else if (strstr(line, ".TS") == line) {
				/* Table start (treated the same as .nf) */
				flush_line(&ctx);
				ctx.printing_table = 1;
				continue;
			} else if (strstr(line, ".TE") == line) {
				/* Table end (treated the same as .fi, but with an extra line break) */
				flush_line(&ctx);
				printf("\n");
				ctx.printing_table = 0;
				continue;
			} else {
				fprintf(stderr, "%s: %s: Found an unrecognized directive on this line: %s\n",
					argv[0], argv[i], line);
				ret = 1;
				goto _cleanup;
			}
		}

		/* If whitespace appears at the start of the line, treat it as a line break
		 * and display the literal whitespace */
		if (*c == ' ' || *c == '\t') {
			flush_line(&ctx);
			print_spaces(ctx.indent);
			ctx.current_x += ctx.indent;

			while (*c == ' ' || *c == '\t') {
				if (*c == '\t') {
					/* Is this the right tab stop? I have no idea! */
					do {
						fputc(' ', stdout);
						ctx.current_x += 1;
					} while (ctx.current_x % 6);
				} else {
					fputc(' ', stdout);
					ctx.current_x += 1;
				}
				c++;
			}
		}

		/* Print it! */
		while (*c) {
			c += process_word(&ctx, c, delimited);
		}

		printf("\033[0m");

		/* When in one of the raw whitespace modes, treat the end of a line
		 * as a forced line break. */
		if (ctx.printing_table) {
			printf("\n");
			ctx.current_x = 0;
			continue;
		}

		/* Should we always be resetting the font to Roman at the end of an input line?
		 * Probably not, but it's worked so far... */
		ctx.current_font = ctx.previous_font = 'R';

		/* Handle macros that want to change indentation after handling one line. */
		if (ctx.next_indent) {
			ctx.indent = ctx.next_indent;
			ctx.next_indent = 0;
			if (ctx.squish_line && ctx.current_x < ctx.indent) {
				ctx.squish_line = 0;
				while (ctx.current_x < ctx.indent) {
					fputc(' ', stdout);
					ctx.current_x += 1;
				}
			} else {
				flush_line(&ctx);
			}
		}
	}

	/* End of file, print the footer. We made it! */
	if (flush_line(&ctx)) printf("\n");
	format_footer(&ctx);

_cleanup:
	if (ctx.topic_title) free(ctx.topic_title);
	if (ctx.topic_section) free(ctx.topic_section);
	if (ctx.topic_date) free(ctx.topic_date);

	if (f != stdin) fclose(f);
	return ret;
}

static int usage(char * argv[]) {
#define X_S "\033[3m"
#define X_E "\033[0m"
	fprintf(stderr,
		"usage: %s [-W " X_S "width" X_E "] " X_S "FILE" X_E "...\n"
		"\n"
		"Parse a very limited subset of ROFF and format it for display on STDOUT, using\n"
		"STDERR as the basis for sizing, suitable for passing to '|more -r'.\n"
		"\n"
		"If an unrecognized directive is found, processing will halt and an error will\n"
		"be printed indicating the unhandled line.\n"
		"\n"
		"Options:\n"
		"\n"
		" -W " X_S "width   Format output for the given width, instead of using the" X_E "\n"
		"            " X_S "width of the terminal." X_E "\n"
		" --help     " X_S "Show this help text." X_E "\n"
		"\n", argv[0]);
	return 1;
}

int main(int argc, char * argv[]) {
	if (isatty(STDERR_FILENO)) ioctl(STDERR_FILENO, TIOCGWINSZ, &w);

	int opt;
	while ((opt = getopt(argc, argv, "?W:-:")) != -1) {
		switch (opt) {
			case 'W':
				w.ws_col = atoi(optarg);
				break;
			case '-':
				if (!strcmp(optarg,"help")) {
					usage(argv);
					return 0;
				}
				fprintf(stderr, "%s: '--%s' is not a recognized long option.\n", argv[0], optarg);
				/* fallthrough */
			case '?':
				return usage(argv);
		}
	}

	if (optind == argc) {
		fprintf(stderr, "%s: no files (see `%s --help` for usage)\n", argv[0], argv[0]);
		return 1;
	}

	int ret = 0;

	for (; optind < argc; optind++) {
		ret |= do_file(argv, optind);
	}

	return ret;
}
